// @angular-builders/jest/dist/schematics/migrations/v21/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.migrateV21 = migrateV21;
const schematics_1 = require("@angular-devkit/schematics");
const utility_1 = require("@schematics/angular/utility");
const json_file_1 = require("@schematics/angular/utility/json-file");
const schematics_2 = require("@angular-builders/common/schematics");
const DEP_BUMPS = {
    jest: '^30.0.0',
    'jest-environment-jsdom': '^30.0.0',
    jsdom: '^26.0.0',
};
function bumpDeps() {
    return (tree) => {
        if (!tree.exists('/package.json'))
            return tree;
        const json = new json_file_1.JSONFile(tree, '/package.json');
        for (const [name, version] of Object.entries(DEP_BUMPS)) {
            if (json.get(['devDependencies', name]) !== undefined) {
                json.modify(['devDependencies', name], version);
            }
            if (json.get(['dependencies', name]) !== undefined) {
                json.modify(['dependencies', name], version);
            }
        }
        return tree;
    };
}
function patchSpecTsconfig() {
    return (0, schematics_2.editJsonFile)('/tsconfig.spec.json', (json) => {
        json.modify(['compilerOptions', 'module'], 'Node16');
        json.modify(['compilerOptions', 'moduleResolution'], 'Node16');
        json.modify(['compilerOptions', 'isolatedModules'], true);
    });
}
const OPTION_RENAMES = {
    configPath: 'config',
    testPathPattern: 'testPathPatterns',
};
function renameBuilderOptions() {
    return (0, utility_1.updateWorkspace)((workspace) => {
        for (const project of workspace.projects.values()) {
            const test = project.targets.get('test');
            if (!test || test.builder !== '@angular-builders/jest:run')
                continue;
            const options = (test.options ?? {});
            for (const [from, to] of Object.entries(OPTION_RENAMES)) {
                if (from in options) {
                    if (!(to in options))
                        options[to] = options[from];
                    delete options[from];
                }
            }
            // Jest 30 renamed the single-string `testPathPattern` to the array-valued
            // `testPathPatterns`. Wrap a carried-over string value so it matches the schema.
            if (typeof options['testPathPatterns'] === 'string') {
                options['testPathPatterns'] = [options['testPathPatterns']];
            }
            test.options = options;
        }
    });
}
const REMOVED_GLOBAL_MOCKS = ['styleTransform', 'getComputedStyle', 'doctype'];
const REMOVED_JEST_OPTIONS = ['browser', 'init', 'mapCoverage', 'testURL', 'timers'];
function stripRemovedOptions() {
    return (0, utility_1.updateWorkspace)((workspace) => {
        for (const project of workspace.projects.values()) {
            const test = project.targets.get('test');
            if (!test || test.builder !== '@angular-builders/jest:run')
                continue;
            const options = (test.options ?? {});
            if (Array.isArray(options['globalMocks'])) {
                options['globalMocks'] = options['globalMocks'].filter((v) => !REMOVED_GLOBAL_MOCKS.includes(v));
            }
            for (const removed of REMOVED_JEST_OPTIONS) {
                if (removed in options)
                    delete options[removed];
            }
            test.options = options;
        }
    });
}
function setZonelessByDetection() {
    return async (tree) => {
        const workspace = await (0, utility_1.readWorkspace)(tree);
        return (0, utility_1.updateWorkspace)((ws) => {
            for (const [name, project] of ws.projects) {
                const test = project.targets.get('test');
                if (!test || test.builder !== '@angular-builders/jest:run')
                    continue;
                if (!(0, schematics_2.isZoneless)(tree, workspace, name)) {
                    const options = (test.options ?? {});
                    options['zoneless'] = false;
                    test.options = options;
                }
            }
        });
    };
}
function migrateV21() {
    return (_tree, context) => {
        context.logger.warn('[@angular-builders/jest] v21 migration applied. Note: tsconfig.spec.json now uses ' +
            'module/moduleResolution "Node16", which may surface pre-existing type errors in your ' +
            'spec code — fix the reported type issues.');
        context.logger.warn('[@angular-builders/jest] Removed globalMocks (styleTransform, getComputedStyle, doctype) ' +
            'were stripped from your config; if your tests relied on them, replace them manually. ' +
            'See MIGRATION.MD (v20→v21) for details.');
        return (0, schematics_1.chain)([
            bumpDeps(),
            patchSpecTsconfig(),
            renameBuilderOptions(),
            stripRemovedOptions(),
            setZonelessByDetection(),
        ]);
    };
}
