// @angular-builders/jest/dist/default-config.resolver.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DefaultConfigResolver = exports.testPattern = void 0;
const lodash_1 = require("lodash");
const core_1 = require("@angular-devkit/core");
const path = __importStar(require("node:path"));
const utils_1 = require("./utils");
const default_config_1 = __importDefault(require("./jest-config/default-config"));
exports.testPattern = `/**/*(*.)@(spec|test).[tj]s?(x)`;
const globalMocks = {
    matchMedia: 'match-media.js',
};
const getMockFiles = (enabledMocks = []) => Object.values((0, lodash_1.pick)(globalMocks, enabledMocks)).map(fileName => (0, core_1.getSystemPath)((0, core_1.normalize)(`${__dirname}/global-mocks/${fileName}`)));
const getSetupFile = (zoneless = true) => {
    const setupFileName = zoneless ? 'setup-zoneless.js' : 'setup-zone.js';
    return (0, core_1.getSystemPath)((0, core_1.normalize)(`${__dirname}/jest-config/${setupFileName}`));
};
class DefaultConfigResolver {
    options;
    // Exposed publicly for testing purposes.
    tsJestTransformRegExp = '^.+\\.(ts|js|mjs|html|svg)$';
    constructor(options) {
        this.options = options;
    }
    resolveGlobal() {
        const setupFilesAfterEnv = [
            getSetupFile(this.options.zoneless ?? true),
            ...getMockFiles(this.options.globalMocks),
        ];
        return { ...default_config_1.default, setupFilesAfterEnv };
    }
    resolveForProject(projectRoot) {
        return {
            testMatch: [`${(0, core_1.getSystemPath)(projectRoot)}${exports.testPattern}`],
            // Scope coverage output to the project directory so that multiple projects
            // in a workspace don't overwrite each other's coverage reports. See #1009.
            coverageDirectory: path.join((0, core_1.getSystemPath)(projectRoot), 'coverage'),
            transform: {
                [this.tsJestTransformRegExp]: [
                    'jest-preset-angular',
                    {
                        // see: jest-preset-angular defaultTransformerOptions https://github.com/thymikee/jest-preset-angular/blob/main/src/presets/index.ts#L11
                        stringifyContentPathRegex: '\\.(html|svg)$',
                        // Join with the default `tsConfigName` if the `tsConfig` option is not provided
                        tsconfig: (0, utils_1.getTsConfigPath)(projectRoot, this.options),
                        // Default to isolatedModules: true for significantly faster compilation.
                        // With isolatedModules: false (the previous implicit default), ts-jest uses the
                        // TypeScript language service to build a full cross-file Program for every test
                        // file, which becomes extremely slow with Angular 19+ code (signals, new control
                        // flow, standalone-by-default). Angular 19+ users report 2min → 15min regressions.
                        // Cross-file type checking is better handled by `tsc --noEmit` or `ng build`.
                        // Users who need the old behaviour can opt out via their jest.config.ts:
                        //   transform: { '...': ['jest-preset-angular', { isolatedModules: false }] }
                        // BREAKING CHANGE: targeted for the next major version.
                        isolatedModules: true,
                    },
                ],
            },
        };
    }
}
exports.DefaultConfigResolver = DefaultConfigResolver;
