// @angular-builders/common/dist/schematics/rules.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.setBuilderForTarget = setBuilderForTarget;
exports.addBuilderDevDependency = addBuilderDevDependency;
exports.removeDevDependencies = removeDevDependencies;
exports.removeFilesIfPresent = removeFilesIfPresent;
exports.editJsonFile = editJsonFile;
const utility_1 = require("@schematics/angular/utility");
const json_file_1 = require("@schematics/angular/utility/json-file");
function setBuilderForTarget(projectName, targetName, builderName, options, opts = {}) {
    return (0, utility_1.updateWorkspace)((workspace) => {
        const project = workspace.projects.get(projectName);
        if (!project)
            throw new Error(`Project "${projectName}" not found.`);
        const target = project.targets.get(targetName);
        if (target) {
            target.builder = builderName;
            // By default merge onto the existing options. `replaceOptions` discards them — used when
            // switching to a builder whose option shape is incompatible with the previous one (e.g.
            // replacing a :unit-test target, whose `runner`/`buildTarget` must not leak to the new builder).
            if (opts.replaceOptions) {
                target.options = (options ?? {});
            }
            else if (options) {
                target.options = { ...(target.options ?? {}), ...options };
            }
        }
        else {
            project.targets.add({ name: targetName, builder: builderName, options: (options ?? {}) });
        }
    });
}
function addBuilderDevDependency(name, version, opts = {}) {
    return (0, utility_1.addDependency)(name, version, {
        type: utility_1.DependencyType.Dev,
        install: opts.install === false ? utility_1.InstallBehavior.None : utility_1.InstallBehavior.Auto,
    });
}
function removeDevDependencies(names) {
    return (tree) => {
        if (!tree.exists('/package.json'))
            return tree;
        const json = new json_file_1.JSONFile(tree, '/package.json');
        for (const name of names) {
            if (json.get(['devDependencies', name]) !== undefined) {
                json.remove(['devDependencies', name]);
            }
        }
        return tree;
    };
}
function removeFilesIfPresent(paths) {
    return (tree) => {
        for (const p of paths) {
            if (tree.exists(p))
                tree.delete(p);
        }
        return tree;
    };
}
function editJsonFile(path, mutator) {
    return (tree) => {
        if (!tree.exists(path))
            return tree;
        const json = new json_file_1.JSONFile(tree, path);
        mutator(json);
        return tree;
    };
}
