// @angular-builders/common/dist/schematics/testing.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.SchematicTestHarness = void 0;
const path = __importStar(require("path"));
const testing_1 = require("@angular-devkit/schematics/testing");
// @schematics/angular's exports map only exposes `.js` files; collection.json
// is not listed so require.resolve('@schematics/angular/collection.json') throws
// in Node 22+ which enforces the exports map. Resolve via package.json instead.
const NG_COLLECTION = path.join(path.dirname(require.resolve('@schematics/angular/package.json')), 'collection.json');
class SchematicTestHarness {
    runner;
    constructor(runner) {
        this.runner = runner ?? new testing_1.SchematicTestRunner('schematics', NG_COLLECTION);
    }
    async createWorkspace(opts = {}) {
        const projects = opts.projects ?? [{ name: 'app' }];
        let tree = await this.runner.runSchematic('workspace', {
            name: 'workspace',
            version: '0.0.0',
            newProjectRoot: 'projects',
        });
        for (const project of projects) {
            tree = await this.runner.runSchematic('application', {
                name: project.name,
                // keep fixtures small + deterministic
                routing: false,
                style: 'css',
                skipTests: false,
                standalone: true,
            }, tree);
        }
        return tree;
    }
}
exports.SchematicTestHarness = SchematicTestHarness;
